### PR TITLE
Merge in relevant upstream commits up to 2022-11-01

### PR DESCRIPTION
**Pull request by FawazTirmizi**
Rebasing or merging everything in is really messy and only a fraction of the commits actually matter to us so I'm just going to cherry-pick what we care about from now on.

Because every commit here touches something we use, I would appreciate some proper reviews instead of rubber-stamp approvals on this, although there shouldn't be anything particularly controversial here.

This only includes commits up to 2022-11-01, there are a couple of bigger changes later on that I think are better suited for separate PRs.

### CODE REVIEW

**Comment by furquan-rivos** [comment]
> Rebasing or merging everything in is really messy and only a fraction of the commits actually matter to us so I'm just going to cherry-pick what we care about from now on.

Is there a lot of effort on merging the commits that don't impact us? I ask because the more we skip, the further we would move away from the upstream oxide repo and it might become difficult in the future to pull in common changes as well.

**Comment by FawazTirmizi** [comment]
> > Rebasing or merging everything in is really messy and only a fraction of the commits actually matter to us so I'm just going to cherry-pick what we care about from now on.
> 
> Is there a lot of effort on merging the commits that don't impact us? I ask because the more we skip, the further we would move away from the upstream oxide repo and it might become difficult in the future to pull in common changes as well.

I was getting a load of conflicts for merge and rebase for some reason, almost all of them being in Oxide's apps or tasks. Also, it fills our history with a bunch of commits completely irrelevant to us (we're currently ~100 commits behind)

I'm going to be sure to pick pretty much anything that touches common code, so I don't think we'll run into too many conflicts there.

**Comment by furquan-rivos** [comment]
> > > Rebasing or merging everything in is really messy and only a fraction of the commits actually matter to us so I'm just going to cherry-pick what we care about from now on.
> > 
> > 
> > Is there a lot of effort on merging the commits that don't impact us? I ask because the more we skip, the further we would move away from the upstream oxide repo and it might become difficult in the future to pull in common changes as well.
> 
> I was getting a load of conflicts for merge and rebase for some reason, almost all of them being in Oxide's apps or tasks. Also, it fills our history with a bunch of commits completely irrelevant to us (we're currently ~100 commits behind)

Why are there merge conflicts in oxide's apps or tasks? Are we touching them at all? If not, then the resolution should be straightforward i.e. always take what they have. I don't think filling history is a problem. I am more concerned about the trees diverging (if our goal is to keep up-to-date with upstream and push changes back)

> 
> I'm going to be sure to pick pretty much anything that touches common code, so I don't think we'll run into too many conflicts there.

I am concerned about larger refactor changes where the PRs would touch common code as well as some of the Oxide task code. In that case we would have to do a larger resolution and not picking the Oxide changes would mean those boards are build-broken in our tree (which could make it harder to detect if any of our common changes are hurting arm platforms).

**Review by beezow — DISMISSED** [review]
Looks fine to me. Seems like mostly API changes with `build_util`, then some timer multiplexing stuff.  I didn't look at any of the oxide apps.

File: build/kconfig/src/lib.rs [hunk 1]
```diff
@@ -0,0 +1,145 @@
...
+
+    /// Regions that tasks have shared access to, keyed by the name the task
+    /// config used to grant access (often peripheral name). These are typically
+    /// memory mapped peripherals.
+    pub shared_regions: BTreeMap<String, RegionConfig>,
+
+    /// Interrupts hooked by the application, keyed by IRQ number.
+    pub irqs: BTreeMap<u32, InterruptConfig>,
+
+    pub timer: (u32, u32),
```
Comment (furquan-rivos) [inline build/kconfig/src/lib.rs:22]: Missing comment. 

Also, this seems to be a Rivos-specific change. Probably add a note to commit message outlining what's changed in this commit compared to what's upstream?

**Review by furquan-rivos — DISMISSED** [review]

File: build/kconfig/src/lib.rs (hunk 1 above)
Comment (FawazTirmizi) [reply build/kconfig/src/lib.rs:22]: Good idea, will do.

**Review by beezow — APPROVED** [review]
Did this include all of the commits from oxide?

**Comment by FawazTirmizi** [comment]
> Did this include all of the commits from oxide?

No, I cherry-picked the ones that were relevant to us, ignoring anything that focused exclusively on their apps. If I recall correctly, there were about a hundred commits upstream between the last rebase we did and 2022-11-01, so this is a fifth of them.

File: build/kconfig/src/lib.rs (hunk 1 above)
Comment (furquan-rivos) [reply build/kconfig/src/lib.rs:22]: Thanks Fawaz. We might want to adopt some process to make it easier for the reviewers to understand which changes are picked as is v/s changes that were adapted. Reference: https://docs.google.com/document/d/1XcFBzz9hLu5ryTEKyMMIO3CcbY-BG-c3otyo4H1uD3E/edit. We can probably use tags like "BACKPORT" and "UPSTREAM" which should make it pretty clear. Anyways, something for the future.

**Review by furquan-rivos — APPROVED** [review]

**Pull request merged**